v0.49.x version of metabase fails to initialize when using mysql 8.0 as database (ApsaraDB / Alibaba Cloud)

### Describe the bug

When I try to use the v0.49.x version of metabase, it fails to initialize

### To Reproduce

Whether upgrading from an older version, or using the new database directly with v0.49.x, the same problem occurs at startup


### Expected behavior

The service is up and running normally.


### Logs

```

2024-03-25 04:11:48,953 INFO plugins.dependencies :: Metabase Vertica Driver dependency {:class com.vertica.jdbc.Driver} satisfied? false


2024-03-25 04:11:48,958 INFO plugins.dependencies :: Plugins with unsatisfied deps: ["Metabase Oracle Driver" "Metabase Vertica Driver"]


2024-03-25 04:11:48,973 INFO metabase.core :: Setting up and migrating Metabase DB. Please sit tight, this may take a minute...


2024-03-25 04:11:48,981 INFO db.setup :: Verifying mysql Database Connection ...


2024-03-25 04:11:49,209 INFO db.setup :: Successfully verified MySQL 8.0.34 application database connection. ✅


2024-03-25 04:11:49,210 INFO db.setup :: Checking if a database downgrade is required...


2024-03-25 04:11:50,261 INFO db.setup :: Running Database Migrations...


2024-03-25 04:11:50,262 INFO db.setup :: Setting up Liquibase...


2024-03-25 04:11:50,578 INFO db.setup :: Liquibase is ready.


2024-03-25 04:11:50,579 INFO db.liquibase :: Checking if Database has unrun migrations...


2024-03-25 04:11:51,644 INFO db.liquibase :: Database has unrun migrations. Checking if migraton lock is taken...


2024-03-25 04:11:51,652 INFO db.liquibase :: No migration lock found.


2024-03-25 04:11:52,133 INFO db.liquibase :: Running 17 migrations ...


2024-03-25 04:11:53,205 ERROR liquibase.changelog :: ChangeSet migrations/001_update_migrations.yaml::v49.00-059::qnkhuat encountered an exception.



UPDATE SUMMARY


Run:                         17


Previously run:             258


Filtered out:                 5


-------------------------------


Total change sets:          280




FILTERED CHANGE SETS SUMMARY


DBMS mismatch:                5



2024-03-25 04:11:53,262 ERROR metabase.core :: Metabase Initialization FAILED


liquibase.exception.CommandExecutionException: liquibase.exception.LiquibaseException: liquibase.exception.MigrationFailedException: Migration failed for changeset migrations/001_update_migrations.yaml::v49.00-059::qnkhuat:


     Reason: clojure.lang.ExceptionInfo: (conn=420) SAVEPOINT 4c7fbe51-8c36-4d93-a6bd-742b2143c2c6 does not exist {:toucan2/context-trace [["resolve connection" {:toucan2.connection/connectable org.mariadb.jdbc.MariaDbConnection}] ["resolve connection" {:toucan2.connection/connectable nil}]]}


	at liquibase.command.CommandScope.execute(CommandScope.java:253)


	at liquibase.Liquibase.lambda$update$0(Liquibase.java:245)


	at liquibase.Scope.lambda$child$0(Scope.java:186)


	at liquibase.Scope.child(Scope.java:195)


	at liquibase.Scope.child(Scope.java:185)


	at liquibase.Scope.child(Scope.java:164)


	at liquibase.Liquibase.runInScope(Liquibase.java:1419)


	at liquibase.Liquibase.update(Liquibase.java:234)


	at liquibase.Liquibase.update(Liquibase.java:212)


	at liquibase.Liquibase.update(Liquibase.java:194)


	at metabase.db.liquibase$migrate_up_if_needed_BANG_.invokeStatic(liquibase.clj:240)


	at metabase.db.liquibase$migrate_up_if_needed_BANG_.invoke(liquibase.clj:222)


	at metabase.db.setup$migrate_BANG_$fn__50987.invoke(setup.clj:80)


	at metabase.db.liquibase$do_with_liquibase$f_STAR___48693.invoke(liquibase.clj:135)


	at metabase.db.liquibase$do_with_liquibase.invokeStatic(liquibase.clj:138)


	at metabase.db.liquibase$do_with_liquibase.invoke(liquibase.clj:126)


	at metabase.db.setup$migrate_BANG_.invokeStatic(setup.clj:75)


	at metabase.db.setup$migrate_BANG_.doInvoke(setup.clj:56)


	at clojure.lang.RestFn.invoke(RestFn.java:445)


	at metabase.db.setup$run_schema_migrations_BANG_.invokeStatic(setup.clj:147)


	at metabase.db.setup$run_schema_migrations_BANG_.invoke(setup.clj:141)


	at metabase.db.setup$setup_db_BANG_$fn__51015$fn__51016.invoke(setup.clj:165)


	at metabase.util.jvm$do_with_us_locale.invokeStatic(jvm.clj:239)


	at metabase.util.jvm$do_with_us_locale.invoke(jvm.clj:225)


	at metabase.db.setup$setup_db_BANG_$fn__51015.invoke(setup.clj:160)


	at metabase.db.setup$setup_db_BANG_.invokeStatic(setup.clj:159)


	at metabase.db.setup$setup_db_BANG_.invoke(setup.clj:153)


	at metabase.db$setup_db_BANG_$fn__51029.invoke(db.clj:69)


	at metabase.db$setup_db_BANG_.invokeStatic(db.clj:64)


	at metabase.db$setup_db_BANG_.invoke(db.clj:55)


	at metabase.core$init_BANG__STAR_.invokeStatic(core.clj:113)


	at metabase.core$init_BANG__STAR_.invoke(core.clj:98)


	at metabase.core$init_BANG_.invokeStatic(core.clj:156)


	at metabase.core$init_BANG_.invoke(core.clj:151)


	at metabase.core$start_normally.invokeStatic(core.clj:168)


	at metabase.core$start_normally.invoke(core.clj:162)


	at metabase.core$entrypoint.invokeStatic(core.clj:201)


	at metabase.core$entrypoint.doInvoke(core.clj:195)


	at clojure.lang.RestFn.invoke(RestFn.java:397)


	at clojure.lang.AFn.applyToHelper(AFn.java:152)


	at clojure.lang.RestFn.applyTo(RestFn.java:132)


	at clojure.lang.Var.applyTo(Var.java:705)


	at clojure.core$apply.invokeStatic(core.clj:667)


	at clojure.core$apply.invoke(core.clj:662)


	at metabase.bootstrap$_main.invokeStatic(bootstrap.clj:31)


	at metabase.bootstrap$_main.doInvoke(bootstrap.clj:28)


	at clojure.lang.RestFn.invoke(RestFn.java:397)


	at clojure.lang.AFn.applyToHelper(AFn.java:152)


	at clojure.lang.RestFn.applyTo(RestFn.java:132)


	at metabase.bootstrap.main(Unknown Source)


Caused by: liquibase.exception.LiquibaseException: liquibase.exception.MigrationFailedException: Migration failed for changeset migrations/001_update_migrations.yaml::v49.00-059::qnkhuat:

....

	... 218 more


2024-03-25 04:11:53,292 INFO metabase.core :: Metabase Shutting Down ...


2024-03-25 04:11:53,299 INFO metabase.server :: Shutting Down Embedded Jetty Webserver


2024-03-25 04:11:53,310 INFO metabase.core :: Metabase Shutdown COMPLETE
```

### Information about your Metabase installation

```JSON
- database: MySQL  8.0.34
- Metabae version : 0.49.0 /0.49.1
- Metabase hosting env:  Docker 
- Metabase internal database:  MySQL  8.0.34
```


### Severity

P1

### Additional context

_No response_

--- a/src/metabase/db/custom_migrations.clj
+++ b/src/metabase/db/custom_migrations.clj
@@ -12,6 +12,7 @@
    [clojure.core.match :refer [match]]
    [clojure.java.io :as io]
    [clojure.set :as set]
+   [clojure.string :as str]
    [clojure.walk :as walk]
    [clojurewerkz.quartzite.jobs :as jobs]
    [clojurewerkz.quartzite.scheduler :as qs]
@@ -90,7 +91,15 @@
 ;; metabase.util/upper-case-en
 (defn- upper-case-en
   ^String [s]
-  (.toUpperCase (str s) Locale/US))
+  (when s
+    (.toUpperCase (str s) Locale/US)))
+
+
+;; metabase.util/lower-case-en
+(defn- lower-case-en
+  ^String [s]
+  (when s
+   (.toLowerCase (str s) Locale/US)))
 
 ;;; +----------------------------------------------------------------------------------------------------------------+
 ;;; |                                                  MIGRATIONS                                                    |
@@ -903,80 +912,29 @@
 (defn- db-type->to-unified-columns
   "Each unified column is 3 items sequence [table-name, column-name, is-nullable?]"
   [db-type]
-  (case db-type
-    :h2      [[:activity :timestamp false]
-              [:application_permissions_revision :created_at false]
-              [:collection_permission_graph_revision :created_at false]
-              [:core_session :created_at false]
-              [:core_user :date_joined false]
-              [:core_user :last_login true]
-              [:core_user :updated_at true]
-              [:dependency :created_at false]
-              [:dimension :created_at false]
-              [:dimension :updated_at false]
-              [:metabase_database :created_at false]
-              [:metabase_database :updated_at false]
-              [:metabase_field :created_at false]
-              [:metabase_field :updated_at false]
-              [:metabase_field :last_analyzed true]
-              [:metabase_fieldvalues :created_at false]
-              [:metabase_table :created_at false]
-              [:metabase_table :updated_at false]
-              [:metric :created_at false]
-              [:metric :updated_at false]
-              [:permissions_revision :created_at false]
-              [:pulse :created_at false]
-              [:pulse :updated_at false]
-              [:pulse_channel :created_at false]
-              [:pulse_channel :updated_at false]
-              [:recent_views :timestamp false]
-              [:report_card :created_at false]
-              [:report_cardfavorite :created_at false]
-              [:report_cardfavorite :updated_at false]
-              [:report_dashboard :created_at false]
-              [:report_dashboard :updated_at false]
-              [:report_dashboardcard :created_at false]
-              [:report_dashboardcard :updated_at false]
-              [:segment :created_at false]
-              [:segment :updated_at false]]
-    :mysql   [[:activity :timestamp false]
-              [:application_permissions_revision :created_at false]
-              [:collection_permission_graph_revision :created_at false]
-              [:core_session :created_at false]
-              [:core_user :date_joined false]
-              [:core_user :last_login true]
-              [:core_user :updated_at true]
-              [:dependency :created_at false]
-              [:dimension :created_at false]
-              [:dimension :updated_at false]
-              [:metabase_field :created_at false]
-              [:metabase_field :last_analyzed true]
-              [:metabase_field :updated_at false]
-              [:metabase_fieldvalues :created_at false]
-              [:metabase_table :created_at false]
-              [:metabase_table :updated_at false]
-              [:metric :created_at false]
-              [:metric :updated_at false]
-              [:permissions_revision :created_at false]
-              [:pulse :created_at false]
-              [:pulse :updated_at false]
-              [:pulse_channel :created_at false]
-              [:pulse_channel :updated_at false]
-              [:recent_views :timestamp false]
-              [:report_card :created_at false]
-              [:report_cardfavorite :created_at false]
-              [:report_cardfavorite :updated_at false]
-              [:report_dashboard :created_at false]
-              [:report_dashboard :updated_at false]
-              [:segment :created_at false]
-              [:segment :updated_at false]]
-   :postgres [[:application_permissions_revision :created_at false]
-              [:collection_permission_graph_revision :created_at false]
-              [:core_user :updated_at true]
-              [:dimension :updated_at false]
-              [:dimension :created_at false]
-              [:permissions_revision :created_at false]
-              [:recent_views :timestamp false]]))
+  (let [query (case db-type
+                :postgres {:select [:table_name :column_name :is_nullable]
+                           :from   [:information_schema.columns]
+                           :where  [:and
+                                    [:= :data_type "timestamp without time zone"]
+                                    [:= :table_schema :%current_schema]
+                                    [:= :table_catalog :%current_database]]}
+
+                :mysql    {:select [:table_name :column_name :is_nullable]
+                           :from   [:information_schema.columns]
+                           :where  [:and
+                                    [:= :data_type "datetime"]
+                                    [:= :table_schema :%database]]}
+                :h2      {:select [:table_name :column_name :is_nullable]
+                          :from   [:information_schema.columns]
+                          :where  [:= :data_type "TIMESTAMP"]})]
+    (->> (t2/query query)
+         (map #(update-vals % (comp keyword lower-case-en)))
+         (remove (fn [{:keys [table_name]}]
+                   (or (#{:databasechangelog :databasechangeloglock} table_name)
+                       (str/starts-with? (name table_name) "v_"))))
+         (map #(update % :is_nullable (fn [x] (= :yes x))))
+         (map (juxt :table_name :column_name :is_nullable)))))
 
 (defn- alter-table-column-type-sql
   [db-type table column ttype nullable?]

--- a/src/metabase/db/custom_migrations.clj
+++ b/src/metabase/db/custom_migrations.clj
@@ -932,6 +932,7 @@
          (map #(update-vals % (comp keyword lower-case-en)))
          (remove (fn [{:keys [table_name]}]
                    (or (#{:databasechangelog :databasechangeloglock} table_name)
+                       ;; excludes views
                        (str/starts-with? (name table_name) "v_"))))
          (map #(update % :is_nullable (fn [x] (= :yes x))))
          (map (juxt :table_name :column_name :is_nullable)))))

